src/security/KeyData.cc

```diff
@@ -151,14 +146,22 @@ Security::KeyData::loadX509PrivateKeyFromFile()
 
 #if USE_OPENSSL
     const char *keyFilename = privateKeyFile.c_str();
+    Ssl::BIO_Pointer bio;
+    if (!Ssl::OpenCertsFileForReading(bio, keyFilename)) {
+        const auto x = ERR_get_error();
+        debugs(83, DBG_IMPORTANT, "ERROR: unable to load key file '" << keyFilename << "': " << ErrorString(x));
+        return false;
+    }
     // XXX: Ssl::AskPasswordCb needs SSL_CTX_set_default_passwd_cb_userdata()
     // so this may not fully work iff Config.Program.ssl_password is set.
     pem_password_cb *cb = ::Config.Program.ssl_password ? &Ssl::AskPasswordCb : nullptr;
-    Ssl::ReadPrivateKeyFromFile(keyFilename, pkey, cb);
-
-    if (pkey && !X509_check_private_key(cert.get(), pkey.get())) {
-        debugs(83, DBG_IMPORTANT, "WARNING: '" << privateKeyFile << "' X509_check_private_key() failed");
-        pkey.reset();
+    while(!pkey && Ssl::ReadPrivateKey(bio, pkey, cb)) {
+        if (pkey && !X509_check_private_key(cert.get(), pkey.get()))
+            pkey.reset(); // check the next available
+    }
+    if (!pkey) {
+        const auto x = ERR_get_error();
+        debugs(83, DBG_IMPORTANT, "ERROR: '" << privateKeyFile << "' failed to load private key: " << ErrorString(x));
```

Comment:
Please refactor to report three different problems separately:

1. Ssl::ReadPrivateKey() failed
2. X509_check_private_key() failed. AFAICT, this is not the same as X509_check_private_key() returning something other than `1`. AFAICT, these errors can be detected via ERR_get_error(), but we will need to flush any stale error for that to work; we may have code that does similar flushing already.
3. bio contained no private keys matching cert. This is not an OpenSSL error -- avoid reporting ERR_get_error().

Reply:
commit a0cc8b0030

Reply:
Commit a0cc8b0 removed a PR-added _loop_ that was searching for a "matching" private key (AFAICT). The official code does not have a loop. I am accepting/resolving this change, but if you think the loop should be there, please note that my change request was compatible with the loop! I am not against this loop if it is useful. On the other hand, perhaps adding that loop is an out-of-scope change for this PR.